File: personal-platform/src/contracts.ts
export const personalScopes = [
  "life:read",
  "calorie:read",
  "calorie:write",
  "kith:read",
  "kith:write",
] as const;

export type PersonalScope = (typeof personalScopes)[number];
export type PersonalApplication = "pace" | "kith";
export type PersonalPlatform = "macos" | "ios";
export type KithEntityType = "person" | "interaction";
export type SyncOperation = "upsert" | "delete";

export interface AuthExchangeInput {
  scopes: PersonalScope[];
  device: {
    id: string;
    application: PersonalApplication;
    platform: PersonalPlatform;
    displayName: string;
  };
}

export interface PersonalIdentity {
  userId: string;
  authUserId: string;
  appleSubject: string;
  email: string | null;
}

export interface AuthExchangeOutput {
  token: string;
  expiresAt: string;
  scopes: PersonalScope[];
  identity: PersonalIdentity;
}

export interface AuthenticatedSession {
  id: string;
  userId: string;
  deviceId: string;
  scopes: PersonalScope[];
  identity: PersonalIdentity;
}

export interface KithPersonPayload {
  id: string;
  name: string;
  howWeMet: string;
  circle: "family" | "close" | "friends" | "work" | "other";
  closeness: number;
  hue: "clay" | "apricot" | "honey" | "rose" | "rust" | "sand" | "sage";
  birthday: string | null;
  standingNotes: string;
  createdAt: string;
  updatedAt: string;
  deletedAt: string | null;
  version: number;
  originDeviceId: string;
}

export interface KithInteractionPayload {
  id: string;
  personId: string;
  kind: "note" | "hangout" | "call" | "message" | "gift" | "milestone" | "remember";
  happenedOn: string;
  body: string;
  createdAt: string;
  updatedAt: string;
  deletedAt: string | null;
  version: number;
  originDeviceId: string;
}

export interface KithMutation {
  idempotencyKey: string;
  entityType: KithEntityType;
  operation: SyncOperation;
  baseVersion: number;
  payload: KithPersonPayload | KithInteractionPayload;
}

export interface KithPushInput {
  mutations: KithMutation[];
}

export interface KithMutationResult {
  idempotencyKey: string;
  entityType: KithEntityType;
  entityId: string;
  status: "accepted" | "duplicate" | "conflict" | "rejected";
  version: number | null;
  sequence: number | null;
  message: string;
  serverRecord?: KithPersonPayload | KithInteractionPayload;
}

export interface KithPushOutput {
  results: KithMutationResult[];
  cursor: number;
  synchronizedAt: string;
}

export interface KithChange {
  sequence: number;
  entityType: KithEntityType;
  entityId: string;
  operation: SyncOperation;
  version: number;
  changedAt: string;
  payload: KithPersonPayload | KithInteractionPayload;
}

export interface KithPullOutput {
  changes: KithChange[];
  cursor: number;
  hasMore: boolean;
  synchronizedAt: string;
}

export interface ToolActionResult {
  actionId: string;
  domain: "calorie" | "kith";
  toolName: string;
  status: "pending" | "succeeded" | "failed" | "undone";
  occurredAt: string | null;
  message: string;
  undo: {
    actionId: string;
    expiresAt: string | null;
  } | null;
}

File: PersonalSyncKit/Sources/PersonalSyncKit/Contracts.swift
import Foundation

public enum PersonalPlatformScope: String, Codable, CaseIterable, Sendable {
    case lifeRead = "life:read"
    case calorieRead = "calorie:read"
    case calorieWrite = "calorie:write"
    case kithRead = "kith:read"
    case kithWrite = "kith:write"
}

public enum PersonalApplication: String, Codable, Sendable {
    case pace
    case kith
}

public enum PersonalClientPlatform: String, Codable, Sendable {
    case macOS = "macos"
    case iOS = "ios"
}

public struct PersonalDevice: Codable, Equatable, Sendable {
    public let id: UUID
    public let application: PersonalApplication
    public let platform: PersonalClientPlatform
    public let displayName: String

    public init(
        id: UUID,
        application: PersonalApplication,
        platform: PersonalClientPlatform,
        displayName: String
    ) {
        self.id = id
        self.application = application
        self.platform = platform
        self.displayName = displayName
    }
}

public struct PersonalIdentity: Codable, Equatable, Sendable {
    public let userId: UUID
    public let authUserId: String
    public let appleSubject: String
    public let email: String?
}

public struct PersonalPlatformSession: Codable, Equatable, Sendable {
    public let token: String
    public let calorieToken: String
    public let expiresAt: Date
    public let scopes: [PersonalPlatformScope]
    public let identity: PersonalIdentity

    public var isExpired: Bool { expiresAt <= Date() }
}

public struct AppleIdentityCredential: Equatable, Sendable {
    public let identityToken: String
    public let nonce: String
    public let email: String?
    public let firstName: String?
    public let lastName: String?

    public init(
        identityToken: String,
        nonce: String,
        email: String?,
        firstName: String?,
        lastName: String?
    ) {
        self.identityToken = identityToken
        self.nonce = nonce
        self.email = email
        self.firstName = firstName
        self.lastName = lastName
    }
}

public struct PersonalSyncConflict: Codable, Equatable, Sendable, Identifiable {
    public let id: UUID
    public let mutationID: UUID
    public let message: String
    public let recordedAt: Date

    public init(
        id: UUID = UUID(),
        mutationID: UUID,
        message: String,
        recordedAt: Date = Date()
    ) {
        self.id = id
        self.mutationID = mutationID
        self.message = message
        self.recordedAt = recordedAt
    }
}

public struct PersonalSyncMutation: Codable, Equatable, Sendable, Identifiable {
    public let id: UUID
    public let idempotencyKey: String
    public let body: Data
    public let createdAt: Date
    public var attemptCount: Int

    public init(
        id: UUID = UUID(),
        idempotencyKey: String,
        body: Data,
        createdAt: Date = Date(),
        attemptCount: Int = 0
    ) {
        self.id = id
        self.idempotencyKey = idempotencyKey
        self.body = body
        self.createdAt = createdAt
        self.attemptCount = attemptCount
    }
}

public struct PersonalSyncSnapshot: Codable, Equatable, Sendable {
    public var cursor: Int
    public var pendingMutations: [PersonalSyncMutation]
    public var conflicts: [PersonalSyncConflict]
    public var lastSynchronizedAt: Date?

    public init(
        cursor: Int = 0,
        pendingMutations: [PersonalSyncMutation] = [],
        conflicts: [PersonalSyncConflict] = [],
        lastSynchronizedAt: Date? = nil
    ) {
        self.cursor = cursor
        self.pendingMutations = pendingMutations
        self.conflicts = conflicts
        self.lastSynchronizedAt = lastSynchronizedAt
    }
}
